src/engine/routing_algorithms/alternative_path_mld.cpp

```diff
@@ -0,0 +1,786 @@
+
+    lhs_cells[0] = get_cell(std::get<0>(lhs.front()));
+    rhs_cells[0] = get_cell(std::get<0>(rhs.front()));
+
+    for (std::size_t i = 0; i < lhs.size(); ++i)
+        lhs_cells[i + 1] = get_cell(std::get<1>(lhs[i]));
+
+    for (std::size_t i = 0; i < rhs.size(); ++i)
+        rhs_cells[i + 1] = get_cell(std::get<1>(rhs[i]));
+
+    std::sort(begin(lhs_cells), end(lhs_cells));
+    std::sort(begin(rhs_cells), end(rhs_cells));
+
+    // Todo: do we need to scale sharing with edge weights in some sort?
+    // Is sharing based on cells only already good enough? Needs experimentation.
+
+    std::size_t num_different = 0;
+
+    // Needs to be wrapped in std::function for Windows compiler bug
+    auto out = boost::make_function_output_iterator(
+        std::function<void(CellID)>([&](auto) { num_different += 1; }));
+
+    std::set_difference(begin(lhs_cells), end(lhs_cells), begin(rhs_cells), end(rhs_cells), out);
+
+    const auto difference = static_cast<double>(num_different) / lhs.size();
+
+    BOOST_ASSERT(difference >= 0.);
+    BOOST_ASSERT(difference <= 1.);
+
+    return 1. - difference;
+}
+
+// Filters packed paths with similar cells compared to the primary route. Mutates range in-place.
+// Returns an iterator to the filtered range's new end.
+template <typename RandIt>
+RandIt filterPackedPathsByCellSharing(const WeightedViaNodePackedPath &path,
+                                      const Partition &partition,
+                                      RandIt first,
+                                      RandIt last)
+{
+    util::static_assert_iter_category<RandIt, std::random_access_iterator_tag>();
+    util::static_assert_iter_value<RandIt, WeightedViaNodePackedPath>();
+
+    const auto over_sharing_limit = [&](const auto &packed) {
+        return normalizedPackedPathSharing(partition, path.path, packed.path) > kGamma;
+    };
+
+    return std::remove_if(first, last, over_sharing_limit);
+}
+
+// Filters packed paths based on local optimality. Mutates range in-place.
+// Returns an iterator to the filtered range's new end.
+template <typename RandIt>
+RandIt filterPackedPathsByLocalOptimality(const WeightedViaNodePackedPath &path,
+                                          const Heap &forward_heap,
+                                          const Heap &reverse_heap,
+                                          RandIt first,
+                                          RandIt last)
+{
+    util::static_assert_iter_category<RandIt, std::random_access_iterator_tag>();
+    util::static_assert_iter_value<RandIt, WeightedViaNodePackedPath>();
+
+    // Check sub-path optimality on alternative path crossing the via node candidate.
+    //
+    //   s - - - v - - - t  our packed path made up of (from, to) edges
+    //        |--|--|       sub-paths "left" and "right" of v based on threshold
+    //        f  v  l       nodes in [v,f] and in [v, l] have to match predecessor in heaps
```

Comment:
without knowing what I explained to you, I would have had difficulties following this explanation